src/libclipper/include/clipper/selection_policies.hpp

```diff
@@ -0,0 +1,287 @@
+    return Derived::hash_models(candidate_models);
+  }
+
+  // Query Pre-processing: select models and generate tasks
+  static std::vector<PredictTask> select_predict_tasks(State state, Query query,
+                                                       long query_id) {
+    return Derived::select_predict_tasks(state, query, query_id);
+  }
+
+  // TODO: change this method name
+  // TODO: I think it may make sense to decouple combine_predictions()
+  // from select_predict_tasks in some cases
+  static Output combine_predictions(State state, Query query,
+                                    std::vector<Output> predictions) {
+    return Derived::combine_predictions(std::forward(state), query,
+                                        predictions);
+  }
+
+  /// When feedback is received, the selection policy can choose
+  /// to schedule both feedback and prediction tasks. Prediction tasks
+  /// can be used to get y_hat for e.g. updating a bandit algorithm,
+  /// while feedback tasks can be used to optionally propogate feedback
+  /// into the model containers.
+  static std::pair<std::vector<PredictTask>, std::vector<FeedbackTask>>
+  select_feedback_tasks(State state, FeedbackQuery query, long query_id) {
+    return Derived::select_feedback_tasks(std::forward(state), query, query_id);
+  }
+
+  /// This method will be called if at least one PredictTask
+  /// was scheduled for this piece of feedback. This method
+  /// is guaranteed to be called sometime after all the predict
+  /// tasks scheduled by `select_feedback_tasks` complete.
+  static State process_feedback(State state, Feedback feedback,
+                                std::vector<Output> predictions) {
+    return Derived::process_feedback(std::forward(state), feedback,
+                                     predictions);
+  }
+
+  static std::string serialize_state(State state) {
+    return Derived::serialize_state(state);
+  }
+
+  static State deserialize_state(const std::string& bytes) {
+    return Derived::deserialize_state(bytes);
+  }
+
+  /**
+   * Human readable debug string for the state
+   */
+  static std::string state_debug_string(const State& state) {
+    return Derived::state_debug_string(state);
+  }
+};
+
+class Exp3Policy : public SelectionPolicy<Exp3Policy, Exp3State> {
+  // Exp3
+  // Select: weighted sampling
+  // Update: update weights based on Loss and respond rate
+
+ public:
+  Exp3Policy() = delete;
+  ~Exp3Policy() = delete;
+  typedef Exp3State state_type;
+
+  constexpr static double eta = 0.01;  // How fast clipper respond to feedback
+
+  static Exp3State initialize(
+      const std::vector<VersionedModelId>& candidate_models);
+
+  static Exp3State add_models(Exp3State state,
+                              const std::vector<VersionedModelId>& new_models);
+
+  static long hash_models(
+      const std::vector<VersionedModelId>& candidate_models) {
+    UNUSED(candidate_models);
```

Comment:
remove unused and comment out the candidate_models argument (`/*candidate_models*/`)